File: mnlc/mnlc_localization/mnlc_global_localization/mnlc_global_costmap_opencv.py
#!/usr/bin/env python

from nav_msgs.msg import OccupancyGrid
from nav_msgs.srv import GetMap
import std_srvs.srv
import numpy as np
import roslib
import rospy
import cv2
import tf

roslib.load_manifest('rbe3002')

class mnlc_global_costmap_opencv():

    def __init__(self):
        self.error = False
        self.flag = 0
        rospy.loginfo("Initializing mnlc_global_costmap_opencv.")
        rospy.init_node("mnlc_global_costmap_opencv")
        self.initialize_params()
        rospy.sleep(self.start_time)
        # give gazebo a chance to warm up so rtabmap doesnt raise an error about not having a map
        rospy.loginfo("mnlc_global_costmap_opencv node ready.")
        self.safe_start()

    def initialize_params(self):
        self.start_time = rospy.get_param('/global_costmap/start_time')
        # grid cost to be considered an obstacle
        # [s] standard service timeout limit
        self.timeout = rospy.get_param('/controller/timeout')
        # number of grid cells to pad the c_scpace with
        self.padding = rospy.get_param('/global_costmap/padding')
        self.first_map = True
        self.map = OccupancyGrid()
        self.listener = tf.TransformListener()

    def safe_start(self):
        rospy.wait_for_service(
            '/rtabmap/get_map', timeout=rospy.Duration(self.timeout))
        try:
            m = rospy.ServiceProxy('/rtabmap/get_map', GetMap)
        except rospy.ServiceException as e:
            rospy.logerr("RTabMap Mappping service call failed: %s" % e)
            self.error_handler()
            return
        rospy.loginfo("RTabMap Mappping service call successful.")
        map = m()
        if map.map.info.resolution == 0:
            rospy.logerr(
                "RTabMap has zero resolution. Exiting...")
            self.error_handler()
            return
        initial_map_metadata = map.map
        self.width = initial_map_metadata.info.width
        self.height = initial_map_metadata.info.height
        self.resolution = initial_map_metadata.info.resolution
        self.gox = initial_map_metadata.info.origin.position.x
        self.goy = initial_map_metadata.info.origin.position.y
        self.rtab_map_pub = rospy.Publisher(
            '/latest_map', OccupancyGrid, queue_size=1)
        self.c_space_pub = rospy.Publisher(
            '/mnlc_global_costmap_opencv/cspace', OccupancyGrid, queue_size=1)
        self.c_space_pub.publish(initial_map_metadata)
        self.rtab_map_sub = rospy.Subscriber(
            '/map', OccupancyGrid, self.update_map, queue_size=1)
        rospy.sleep(self.timeout)
        rospy.wait_for_service(
            '/begin_phase1', timeout=rospy.Duration(self.timeout))
        try:
            tmp = rospy.ServiceProxy('/begin_phase1', std_srvs.srv.Empty)
        except rospy.ServiceException as e:
            rospy.logerr("Begin phase1 service call failed: %s" % e)
            self.error_handler()
            return
        rospy.loginfo("Begin phase1 service call successful.")
        temp = tmp()
        self.listener.waitForTransform(
            '/odom', '/base_footprint', rospy.Time(0), timeout=rospy.Duration(self.timeout))
        # allocate a thread for publishing
        self.calculate_global_costmap()

    def calculate_global_costmap(self):
        cspace = OccupancyGrid()
        cspace.header.frame_id = '/map'
        cspace.info.origin.position.x = self.gox
        cspace.info.origin.position.y = self.goy
        cspace.info.resolution = self.resolution
        cspace.info.height = self.height
        cspace.info.width = self.width
        kernel1 = np.ones((self.padding, self.padding), np.float)
        while not rospy.is_shutdown():
            # time_init = rospy.get_time()
            img = self.img
            unkown_indices = self.unkown_indices
            img_dilate = cv2.dilate(img, kernel=kernel1, iterations=1)
            gaussian_blur = cv2.GaussianBlur(
                src=img_dilate, ksize=(7, 7), sigmaX=2, sigmaY=2)
            norm_image = cv2.normalize(
                gaussian_blur, None, alpha=0, beta=100, norm_type=cv2.NORM_MINMAX, dtype=cv2.CV_32F)
            dataFromGridC = norm_image.flatten('C')
            dataFromGridC[unkown_indices] = -1
            cspace.header.stamp = rospy.Time.now()
            cspace.data = dataFromGridC
            self.c_space_pub.publish(cspace)
            # time_end = rospy.get_time()
            # print("Calculating Global CSpace took: ", time_end - time_init, ".")

    def update_map(self, map):
        path = r'/home/quant/.ros/global_costmap.pgm'
        self.rtab_map_pub.publish(map)
        mapdata_asarray = np.array(object=map.data, copy=False)
        self.unkown_indices = np.where(mapdata_asarray == -1)
        mapdata_asarray[mapdata_asarray == -1] = 0
        byte_array = bytearray(np.uint8(mapdata_asarray).tolist())
        with open('global_costmap.pgm', 'wb') as f:
            f.write('P5' + '\n')
            f.write(str(self.width) + ' ' +
                    str(self.height) + '\n')
            f.write(str(255) + '\n')
            f.write(byte_array)
            f.close()
        self.img = cv2.imread(path, -1)

    def error_handler(self):
        self.error = True

    def run(self):
        while not rospy.is_shutdown() and self.error == False:
            rospy.spin()


if __name__ == '__main__':
    try:
        mnlc_global_costmap_opencv().run()
    except rospy.ROSInterruptException:
        pass

File: src/mnlc_global_costmap_opencv.cpp
#include "mnlc_global_costmap_opencv.h"

void update_map(const nav_msgs::OccupancyGrid::ConstPtr &map)
{
  ros::Time time_init = ros::Time::now();
  double ti = time_init.toSec();
  nav_msgs::OccupancyGrid m = *map;
  mapdata = m;
  frontiers.clear();
  cv::Mat frontier_image = cv::Mat::ones(m.info.width, m.info.height, CV_8UC1);
  std::unordered_map<int, int> uk;
  int index = 0;
  int row = 0;
  int col = 0;
  std::vector<int8_t, std::allocator<int8_t>> temp = m.data;
  BOOST_FOREACH (int8_t &datapoint, temp)
  {
    row = index / m.info.width;
    col = index % m.info.width;
    // printf("row: %d \t", row);
    // printf("col: %d \n", col);
    if (datapoint == 100)
    {
      frontier_image.ptr<unsigned char>(col)[row] = 0;
    }
    else if (datapoint == 0)
    {
      frontier_image.ptr<unsigned char>(col)[row] = 255;
    }
    else if (datapoint == -1)
    {
      uk.emplace(index, 0);
      m.data[index] = 0;
      frontier_image.ptr<unsigned char>(col)[row] = 205;
    }
    unsigned char ch = (unsigned char)m.data[index];
    image.ptr<unsigned char>(col)[row] = ch;
    index++;
  }
  // for (auto itr = uk.begin(); itr != uk.end(); itr++)
  //       std::cout << itr->first << "\t" << itr->second << std::endl;
  unknown_indices = uk;
  cv::Mat original;
  cv::inRange(frontier_image, 0, 1, original);
  cv::Mat edges;
  const double threshold1 = 0.0;
  const double threshold2 = 255.0;
  const int kernel_size = 3;
  cv::Canny(frontier_image, edges, threshold1, threshold2, kernel_size, true);
  std::vector<std::vector<cv::Point>> contours1;
  cv::findContours(original, contours1, cv::RETR_LIST, cv::CHAIN_APPROX_TC89_L1);
  cv::drawContours(original, contours1, -1, 255, 5);
  cv::Mat tmp;
  cv::bitwise_not(original, tmp);
  cv::Mat frontier;
  cv::bitwise_and(tmp, edges, frontier);
  std::vector<std::vector<cv::Point>> contours2;
  cv::findContours(frontier, contours2, cv::RETR_LIST, cv::CHAIN_APPROX_TC89_L1);
  cv::drawContours(frontier, contours2, -1, 255, 2);
  std::vector<std::vector<cv::Point>> contours3;
  cv::findContours(frontier, contours3, cv::RETR_LIST, cv::CHAIN_APPROX_TC89_L1);
  float res = m.info.resolution;
  float gox = m.info.origin.position.x;
  float goy = m.info.origin.position.y;
  cv::Point2f p;
  cv::Moments moment;
  if (sizeof(contours3) > 0)
  {
    BOOST_FOREACH (std::vector<cv::Point> &contour, contours3)
    {
      moment = cv::moments(contour, true);
      int cx = (int)(moment.m10 / moment.m00);
      int cy = (int)(moment.m01 / moment.m00);
      float xr = cx * res + gox;
      float yr = cy * res + goy;
      // printf("xr: %f\t yr: %f\n", xr, yr);
      p.x = xr;
      p.y = yr;
      frontiers.push_back(p);
    }
  }
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "mnlc_global_costmap_opencv");
  ros::NodeHandle n;
  double start_time;
  n.getParam("/global_costmap/start_time", start_time);
  double timeout;
  n.getParam("/controller/timeout", timeout);
  int padding;
  n.getParam("/global_costmap/padding", padding);
  bool first_map = true;
  int detector_padding = (int)(padding / 2);
  visualization_msgs::Marker points;
  points.header.frame_id = "/map";
  points.header.stamp = ros::Time::now();
  points.ns = "markers";
  points.id = 9;
  points.type = visualization_msgs::Marker::POINTS;
  points.action = visualization_msgs::Marker::ADD;
  points.pose.orientation.w = 1.0;
  points.scale.x = points.scale.y = 0.05;
  points.color.r = 255.0 / 255.0;
  points.color.g = 0.0 / 255.0;
  points.color.b = 0.0 / 255.0;
  points.color.a = 1;
  points.lifetime = ros::Duration();
  nav_msgs::OccupancyGrid map;
  tf::TransformListener listener;
  ros::Time st;
  st.fromSec(start_time);
  ros::Time::sleepUntil(st);
  ros::service::waitForService("/rtabmap/get_map", ros::Duration(timeout));
  ros::ServiceClient client1 = n.serviceClient<const nav_msgs::GetMap>("/rtabmap/get_map");
  nav_msgs::GetMap srv1;
  if (client1.call(srv1))
  {
    ROS_INFO("RTabMap Mapping service call successful.");
  }
  else
  {
    ROS_ERROR("RTabMap Mapping service call failed.");
    exit(0);
  }
  if (srv1.response.map.info.resolution == 0)
  {
    ROS_ERROR("RTabMap has zero resolution. Exiting...");
    exit(0);
  }
  nav_msgs::OccupancyGrid initial_map_metadata = srv1.response.map;
  int width = initial_map_metadata.info.width;
  int height = initial_map_metadata.info.height;
  double resolution = initial_map_metadata.info.resolution;
  double gox = initial_map_metadata.info.origin.position.x;
  double goy = initial_map_metadata.info.origin.position.y;
  ros::Publisher rtab_map_pub = n.advertise<nav_msgs::OccupancyGrid>("/latest_map", 1);
  ros::Publisher cspace_pub = n.advertise<nav_msgs::OccupancyGrid>("/mnlc_global_costmap_opencv/cspace", 1);
  ros::Publisher detected_opencv_pub = n.advertise<geometry_msgs::PointStamped>("/opencv_points", 10);
  ros::Publisher detected_opencv_arr_pub = n.advertise<PointArray>("/opencv_points_arr", 1);
  ros::Publisher shapes_pub = n.advertise<visualization_msgs::Marker>("/OpenCVFrontierDetector/shapes", 10);
  cspace_pub.publish(initial_map_metadata);
  image = cv::Mat::ones(width, height, CV_8UC1);
  ros::Subscriber rtabmap_sub = n.subscribe("/map", 10, update_map);
  ros::Time sr;
  sr.fromSec(timeout);
  ros::Time::sleepUntil(sr);
  ros::spinOnce();
  ros::service::waitForService("/begin_phase1", ros::Duration(timeout));
  ros::ServiceClient client2 = n.serviceClient<const std_srvs::Empty>("/begin_phase1");
  std_srvs::Empty srv2;
  if (client2.call(srv2))
  {
    ROS_INFO("Begin phase1 service call successful.");
  }
  else
  {
    ROS_ERROR("Begin phase1 service call failed:");
    exit(0);
  }
  listener.waitForTransform("/odom", "/base_footprint", ros::Time(0), ros::Duration(timeout));
  nav_msgs::OccupancyGrid cspace;
  cspace.header.frame_id = "/map";
  cspace.header.stamp = ros::Time::now();
  cspace.info.width = width;
  cspace.info.height = height;
  cspace.info.resolution = resolution;
  cspace.info.origin.position.x = gox;
  cspace.info.origin.position.y = goy;
  int dilation_type = cv::MORPH_RECT;
  int dilation_size = padding;
  cv::Mat kernel = cv::getStructuringElement(dilation_type,
                                             cv::Size(dilation_size, dilation_size),
                                             cv::Point((int)(dilation_size / 2), (int)(dilation_size / 2)));
  geometry_msgs::PointStamped exploration_goal;
  exploration_goal.header.frame_id = "/map";
  exploration_goal.point.z = 0;
  PointArray point_array;
  ros::Rate loop_rate(60);
  while (ros::ok())
  {
    std::unordered_map<int, int> uk = unknown_indices;
    cv::Mat img_dilate;
    cv::Mat img = image;
    cv::dilate(img, img_dilate, kernel);
    cv::Mat gaussian_bur;
    cv::Size size;
    size.width, size.height = 7;
    cv::GaussianBlur(img_dilate, gaussian_bur, size, 1.0, 1.0);
    cv::Mat norm_image;
    cv::normalize(gaussian_bur, norm_image, 0, 100, cv::NormTypes::NORM_MINMAX);
    std::vector<int8_t, std::allocator<int8_t>> data_from_grid_c;
    for (int i = 0; i < height; i++)
    {
      for (int j = 0; j < width; j++)
      {
        auto it = uk.find(i * width + j);
        if (it != uk.end())
        {
          data_from_grid_c.push_back(-1);
        }
        else
        {
          data_from_grid_c.push_back(norm_image.ptr<unsigned char>(j)[i]);
        }
      }
    }
    point_array.points.clear();
    BOOST_FOREACH (cv::Point2f &frontier, frontiers)
    {
      exploration_goal.header.stamp = ros::Time(0);
      exploration_goal.point.x = frontier.y;
      exploration_goal.point.y = frontier.x;
      detected_opencv_pub.publish(exploration_goal);
      point_array.points.push_back(exploration_goal);
      geometry_msgs::Point p;
      p.x = exploration_goal.point.x;
      p.y = exploration_goal.point.y;
      points.points.push_back(p);
    }
    detected_opencv_arr_pub.publish(point_array);
    shapes_pub.publish(points);
    points.points.clear();
    cspace.header.stamp = ros::Time::now();
    mapdata.header.stamp = ros::Time::now();
    cspace.data = data_from_grid_c;
    cspace_pub.publish(cspace);
    rtab_map_pub.publish(mapdata);
    ros::spinOnce();
    loop_rate.sleep();
  }
  return 0;
}